File: pfc.pl
#!/usr/bin/perl

use strict;
use warnings;
use Cwd;
use Getopt::Long;
use Digest::SHA qw(sha1_hex);
use 5.010;

my $HASH = 0;
my $DEBUG = 0;

GetOptions (
    'h|hash' => \$HASH,
    'debug' => \$DEBUG,
);

my @paths;
if (@ARGV) {
    foreach my $input_path (@ARGV) {
        if (-d $input_path) {
            $input_path =~ s{(?<!/)$}{/};
            push @paths, $input_path;
        }
    }
}
else {
    push @paths, getcwd().'/';
}

my %hash_files;
foreach my $path (@paths) {
    foreach my $file (map($path.$_, file_list($path))) {
        my $sha = Digest::SHA->new();

        open my $FH, $file or next;
        $sha->addfile($FH);
        my $hash_code = $sha->hexdigest;
        close $FH;

        push @{$hash_files{$hash_code}}, $file;
    }
}

my %results;
foreach my $file_hash (keys %hash_files) {
    if (@{$hash_files{$file_hash}} > 1) {
        my @sorted_files = sort @{$hash_files{$file_hash}};
        @{$results{$file_hash}} = (@sorted_files);
    }
}

my @hash_list = sort { $results{$a}->[0] cmp $results{$b}->[0] } keys %results;

foreach my $hash (@hash_list) {
    print "$hash\n" if $HASH;
    print join("\n", (@{$results{$hash}}))."\n\n";
}

sub file_list {
    my $path = shift;

    my @dirs = ($path);
    my @file_list;
    my ($dir, $file);
    while ($dir = pop(@dirs)) {
        local *DH;

        if (!opendir(DH, $dir)) {
            warn "Cannot opendir $dir: $!";
            next;
        }

        foreach (readdir(DH)) {
            $file = $dir.$_;

            next if ($_ eq '.' || $_ eq '..' || -l $file);

            if (-d $file) {
                $file .= '/';
                push @dirs, $file ;
            }
            else {
                my $rel_file = $file;
                $rel_file =~ s/$path//;
                push @file_list, $rel_file;
            }
        }
        closedir(DH);
    }

    @file_list;
}

File: pfc.py
#!/usr/bin/python

import sys
import getopt
import os
import time
import stat
import hashlib

HASH = False
DEBUG = 0

def usage():
    print """
USAGE:
    python pfc.py [-h|--hash] [-d|--debug]
"""

def init():
    try:
        opts, args = getopt.getopt(sys.argv[1:], 'hd', ['hash', 'debug='])
    except getopt.GetoptError:
        usage()
        sys.exit()

    global HASH
    global DEBUG

    for opt, arg in opts:
        if opt == '--hash' or opt == '-h':
            HASH = True
        if opt == '-d':
            DEBUG = 1
        if opt == '--debug':
            DEBUG = int(arg)

    if not args:
        args.append(os.getcwd() + os.path.sep)

    return args

def debug(msg, level = 1):
    if level <= DEBUG:
        ori_time = time.time();
        crt_time = (time.strftime('%H:%M:%S',time.localtime(ori_time))
                 + '.' + str(ori_time - int(ori_time))[2:5])
        print "[DebugMessage %s %d] %s" % (crt_time, level, msg)

def getFileList(path):
    if os.path.exists(path):
        dirlist = []

        for name in os.listdir(path):
            fullname = os.path.join(path, name)
            st = os.lstat(fullname)
            if stat.S_ISDIR(st.st_mode):
                dirlist += getFileList(fullname)
            elif stat.S_ISREG(st.st_mode):
                    dirlist.append(fullname)

        return dirlist
    else:
        return []

def fileSHA1(filepath):
    with open(filepath,'rb') as f:
        sha1obj = hashlib.sha1()
        sha1obj.update(f.read())
        hash = sha1obj.hexdigest()
        return hash

if __name__ == '__main__':
    hash_files = {}

    for path in init():
        debug(path, 1)
        for filepath in getFileList(path):
            debug(filepath, 2)
            fSHA1 = fileSHA1(filepath)
            debug(fSHA1, 3)

            if fSHA1 not in hash_files.keys():
                hash_files[fSHA1] = []

            hash_files[fSHA1].append(filepath)

    results = {}
    for hash_code in hash_files.keys():
        if hash_files[hash_code][1:]:
            results[hash_code] = sorted(hash_files[hash_code])

    for group in sorted(results.iteritems(), key = lambda x: x[1][0]):
        if HASH: print group[0]
        for f in group[1]: print f
        print
